R/bookmark-state.R

```diff
@@ -0,0 +1,960 @@
+    # some special handling.
+    input = NULL,
+
+    # Directory for extra files, if restoring from state that was saved to disk.
+    dir = NULL,
+
+    # For values other than input values. These values don't need the special
+    # phandling that's needed for input values, because they're only accessed
+    # from the onRestore function.
+    values = NULL,
+
+    initialize = function(queryString = NULL) {
+      self$reset() # Need this to initialize self$input
+
+      if (!is.null(queryString) && nzchar(queryString)) {
+        tryCatch(
+          withLogErrors({
+            qsValues <- parseQueryString(queryString, nested = TRUE)
+
+            if (!is.null(qsValues[["__subapp__"]]) && qsValues[["__subapp__"]] == 1) {
+              # Ignore subapps in shiny docs
+              self$reset()
+
+            } else if (!is.null(qsValues[["_state_id_"]]) && nzchar(qsValues[["_state_id_"]])) {
+              # If we have a "_state_id_" key, restore from saved state and
+              # ignore other key/value pairs. If not, restore from key/value
+              # pairs in the query string.
+              self$active <- TRUE
+              private$loadStateQueryString(queryString)
+
+            } else {
+              # The query string contains the saved keys and values
+              self$active <- TRUE
+              private$decodeStateQueryString(queryString)
+            }
+          }),
+          error = function(e) {
+            # If there's an error in restoring problem, just reset these values
+            self$reset()
+            self$initErrorMessage <- e$message
+            warning(e$message)
+          }
+        )
+      }
+    },
+
+    reset = function() {
+      self$active <- FALSE
+      self$initErrorMessage <- NULL
+      self$input <- RestoreInputSet$new(list())
+      self$values <- new.env(parent = emptyenv())
+      self$dir <- NULL
+    },
+
+    # This should be called before a restore context is popped off the stack.
+    flushPending = function() {
+      self$input$flushPending()
+    },
+
+
+    # Returns a list representation of the RestoreContext object. This is passed
+    # to the app author's onRestore function. An important difference between
+    # the RestoreContext object and the list is that the former's `input` field
+    # is a RestoreInputSet object, while the latter's `input` field is just a
+    # list.
+    asList = function() {
+      list(
+        input = self$input$asList(),
+        dir = self$dir,
+        values = self$values
+      )
+    }
+  ),
+
+  private = list(
+    # Given a query string with a _state_id_, load saved state with that ID.
+    loadStateQueryString = function(queryString) {
+      values <- parseQueryString(queryString, nested = TRUE)
+      id <- values[["_state_id_"]]
+
+      # This function is passed to the loadInterface function; given a
+      # directory, it will load state from that directory
+      loadFun <- function(stateDir) {
```

Comment:
For some common error cases, can we throw explicit errors (since they will be displayed to the user)? stateDir doesn't exist, missing files that we expect, etc.